### ISSUE TITLE
Performance degradation during batched inserts

### ISSUE THREAD

**Issue by jbennettgit** [terminusdb/terminusdb#253]
Ubuntu 18.04 running bootstrap version 4.1.2 in docker container, using python client.  Time to insert a polity's data from a test version of a seshat csv increases linearly even if the number of inserts (batched by polity) stays roughly the same.  Attempts to squash() after each execute() commit only increased the incremental and total time.

Separately attempts to view the final, committed db in the TerminusDB console hangs with a spinning icon.

To reproduce:

1. Pull our seshat test code using git clone https://github.com/MajidBenam/seshat-3store.git 
2. cd seshat-3store
3. cp -f equinox_full.csv equinox.csv
4. python3 define_seshat_schema.py
5. python3 insert_from_csv.py

Step 3 ensures you use the full 47K line csv file, not the little test equinox.csv.
Step 4 should take around 7-8s to define the schema.

During step 5 the initial set of commits, which batch insertions by polity, show ~450 inserts happening in ~1s.  But after 20 or so polity commits things really start to slow and by the end each polity takes ~10s to commit.  We expect roughly constant time per insert.

- If you re-run step 4, it will delete the db and rebuild it (and the schema) for easy testing.
- There will be various complaints and WARNINGs during step 5 which you can ignore.
- If you run step 5 once and then run step 5 immediately again (re-asserting the same data so causing deletes as well as inserts) quite often there is an initial 'hang' of a few 10s of seconds at the beginning during the commit of first true polity.

Partial example output from step 5 enclosed below.  Each 'Committing' line reports the number of inserts (i:), deletes (d:) and elapsed time for the execute() call.

Parsing data from equinox.csv
Committing data for Code book i:6 d:0 0.15s
WARNING: Pre-casting '1790000:490000' from AfDurrn|Social Complexity variables|Social Scale|Polity territory to a range: likely ill-formed!
Committing data for AfDurrn i:453 d:0 0.92s
WARNING: Pre-casting '60000:80000' from AfGhurd|Social Complexity variables|Social Scale|Polity territory to a range: likely ill-formed!
WARNING: Pre-casting '900000:1100000' from AfGhurd|Social Complexity variables|Social Scale|Polity territory to a range: likely ill-formed!
WARNING: Pre-casting '1600000:1700000' from AfGhurd|Social Complexity variables|Social Scale|Polity territory to a range: likely ill-formed!
Committing data for AfGhurd i:485 d:0 0.92s
WARNING: Pre-casting '600000:800000' from AfGrBct|Social Complexity variables|Social Scale|Polity territory to a range: likely ill-formed!
WARNING: Pre-casting '1500000:2000000' from AfGrBct|Social Complexity variables|Social Scale|Polity Population to a range: likely ill-formed!
WARNING: Pre-casting '25000:50000' from AfGrBct|Social Complexity variables|Social Scale|Population of the largest settlement to a range: likely ill-formed!
Committing data for AfGrBct i:506 d:0 1.00s
WARNING: Pre-casting '2750000:3000000' from AfHepht|Social Complexity variables|Social Scale|Polity territory to a range: likely ill-formed!
Committing data for AfHepht i:568 d:0 1.11s

....

Committing data for YeQasmi i:414 d:0 13.88s
WARNING: Pre-casting '30000:40000' from YeQatab|Social Complexity variables|Social Scale|Population of the largest settlement to a range: likely ill-formed!
Committing data for YeQatab i:300 d:0 9.37s
WARNING: Unable to pre-cast 'suspected unknown' from YeRasul|Warfare variables|Military Technologies|Long walls to integer; assuming 0
Committing data for YeRasul i:441 d:0 12.71s
Committing data for YeSaRay i:222 d:0 7.17s
WARNING: Pre-casting '30000:40000' from YeSabaC|Social Complexity variables|Social Scale|Population of the largest settlement to a range: likely ill-formed!
Committing data for YeSabaC i:279 d:0 9.17s
WARNING: Unable to pre-cast 'suspected unknown' from YeTahir|Social Complexity variables|Social Scale|Polity territory to decimal; assuming 0.0
WARNING: Unable to pre-cast 'suspected unknown' from YeTahir|Social Complexity variables|Social Scale|Polity Population to decimal; assuming 0.0
WARNING: Unable to pre-cast 'suspected unknown' from YeTahir|Social Complexity variables|Social Scale|Population of the largest settlement to decimal; assuming 0.0
WARNING: Unable to pre-cast 'suspected unknown' from YeTahir|Warfare variables|Military Technologies|Long walls to integer; assuming 0
Committing data for YeTahir i:435 d:0 13.29s
WARNING: Pre-casting '250000:350000' from YeWarLd|Social Complexity variables|Social Scale|Polity territory to a range: likely ill-formed!
WARNING: Unable to pre-cast 'suspected unknown' from YeWarLd|Warfare variables|Military Technologies|Long walls to integer; assuming 0
Committing data for YeWarLd i:456 d:0 12.78s
WARNING: Unable to pre-cast 'suspected unknown' from YeZiyad|Warfare variables|Military Technologies|Long walls to integer; assuming 0
Committing data for YeZiyad i:402 d:0 11.15s
Total assertions: 374 requiring inserts: 169681 deletes: 0
Total failed commits: 0
Execution time: 2499.8s

**Comment by rrooij** [comment]
I see that you are trying to insert on each row. It is better to do a bulk insert with: `post`, `remote` or `file`. `post` will send the CSV to the server, `remote` fetches it from a remote source and `file` will search for a local file on the same system that TerminusDB is running on.

An example can be found here: https://github.com/terminusdb/terminusdb-tutorials/blob/master/bike-tutorial/python/bike-tutorial.py#L49

**Comment by jbennettgit** [comment]
I am familiar with that example.  However while the data source is a spreadsheet the data structures being asserted are not simple (as Kevin and Gavin know) and different numbers of inserts (and deletes) are made depending on the information on each row.  Further the values of many properties are class instances themselves and have to be managed.  So we cannot use the csv support that comes native from TerminusDB (and eventually we'll be using a different front end source for the data in any case).  

An original version tried to make a set of inserts and deletes per spreadsheet line but that was really slow (it would have taken over 2 days for the full spreadsheet).  This version attempts to batch all the inserts and deletes according to each 'polity' which is a natural unit for our application. This reduced the time for the full spreadsheet to <1hr to make nearly 170K inserts over 374 polities. If each polity commit took around 1s (as it does at the start) then I expect the whole operation to take around 6m.

I suspect, with no evidence, that the problem is the following.  In our application we will be asserting  possibly several triples for a property on a polity.  When we are updating that property with its new values we need to ensure we delete the old triples first (so we don't have residual, possibly incorrect data left over).  So we do something like (assume existing polity 'AfDurn'):

triple('v:Polity_ID','original_Name','Afdurn').
triple('v:Polity_ID','Polity territory','v:Old_Values_1').
delete_triple('v:Polity_ID','Polity territory','v:Old_Values_1').
delete_object('v:Old_Values_1').
...code to insert new Polity territory triples on v:Polity_ID..

The 2nd triple() finds the previous Polity territory values, the delete_triple() gets rid of them, and the delete_object() gets rid of the now disconnected _Values instances that held on to the old values (so we don't clutter the DB and don't inadvertently have naming collisions with otherwise anonymous Value instances).  I think what is happening is that as more triples are added for all the other polities the 2nd triple might just take longer and longer to do the match (unless there is an index on Polity_ID).  We do this idiom once for each property on the polity we are updating (~20 properties or so per execute, each with their own v:Old_Values_n variable).

If there is a better way, let us know. thanks.

**Comment by GavinMendelGleason** [comment]
Hello Jim, I'll look at this this weekend.

**Comment by jbennettgit** [comment]
Thanks Gavin!  Let me know if you have any questions.

BTW I did some more experiments and gathered more timing data.

As I mentioned in my report follow-on we ‘batch’ all property updates on a per polity basis to reduce the calls to execute().  The structure of each polity query over all looks like:

qp,[qf,qv+]+  

where 

- qp is the fragment of the query that either looks up the existing polity object or creates it (via insert(), etc.)
- qf is the fragment that finds and deletes the old triples (and the scoped objects if needed) for the property
- qv are the query fragments that assert the new value(s) for the property

I attempted to isolate the location(s) of the timing issues. Each time I ran the experiment I did the following thing:

terminusdb-container stop
terminusdb-container rm (deleting the full store)
terminusdb-container run
python3 define_seshat_schema.py

So for each experiment the database had nothing in it. As a consequence I could:

1. Just run the qp fragments: timing starts at 0.14s and grows to 0.5s after 374 queries. I assume the increased timing reflects the work needed to both create and index the object(s) plus journaling.
2. Run a modified qp only that just does the insert without looking to see if the polity already exists: Same timing result as experiment 1.

Of course in both there are only very few triples in the db (374 objects and 374 triples) but the timing trend is clear.  

3. Run qp,[qv]+, that is, drop the qf flushing fragments (since I knew the db was clean, nothing to flush).  Minor timing improvement but calls were still taking 10s at the end.  So my previous conjecture about it being the flushing lookup code looks incorrect and it is something about the qv inserts.

4. Run qp then run qp,[qf,qv+]+ that is do the probe/insert for the polity twice in succession (as separate execute() calls).  I expected the 2nd qp to simply match the results of the first qp, which did the creation. In this case, the timings on the first qp increased from 0.14s to 0.76 (1/2 again as much as experiment 1, probably consistent with also many more triples for prior qvs being in the db).  But the timings on the 2nd query doubled(!) so that by the end they were taking ~20s!  I have no idea why running the first qp should make the second full query take even longer.
